homekit/StateManager.js

```diff
@@ -1,623 +1,1039 @@
+				temperatureUnit: device.temperatureUnit,
+				mode: device.state.mode,
+				fanSpeed: device.state.fanSpeed,
+				swing: device.state.verticalSwing,
+				horizontalSwing: device.state.horizontalSwing,
+				light: device.state.light ? 'on' : 'off'
+			}
 
-// TODO: perhaps make this a class?
-module.exports = (device, platform) => {
-	Characteristic = platform.api.hap.Characteristic
-	const log = platform.log
-	const enableClimateReactAutoSetup = platform.enableClimateReactAutoSetup
+			smartModeState.lowTemperatureThreshold = device.state.targetTemperature - (device.usesFahrenheit ? 1.8 : 1)*climateReactAutoSetupMultiplier + climateReactAutoSetupOffset
+			smartModeState.lowTemperatureState = {
+				on: true,
+				targetTemperature: device.state.targetTemperature,
+				temperatureUnit: device.temperatureUnit,
+				mode: device.state.mode,
+				fanSpeed: device.state.fanSpeed,
+				swing: device.state.verticalSwing,
+				horizontalSwing: device.state.horizontalSwing,
+				light: device.state.light ? 'on' : 'off'
+			}
+		}
+
+		// StateHandler is invoked as a Proxy, and therefore overwrites/intercepts the default get()/set() commands [traps]
+		// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy
+
+		// NOTE: device.state is of "type" StateHandler. When one of its properties is "set" (e.g. device.state.<property> = <val>),
+		//       that's where we actually send commands to the appropriate Sensibo devices. If a property is not set, the aformentioned
+		//       code will not execute and the changes would not take effect.
+		//
+		//       For example, if we set a property of smartMode directly, e.g. device.state.smartMode.enabled = true, StateHandler's
+		//       setter will not get called and so any changes will not take effect. This is why we MUST update a device's property as
+		//       a whole, and do it only once (otherwise's the setter will get called multiple times which will send repeated commands
+		//       to the Sensibo devices).
+		device.state.smartMode = smartModeState
+	}
 
 	return {
 
 		get: {
 			// TODO: refactor this similar to PureActive below?
+			/** @param {homebridge.CharacteristicGetCallback} callback */
 			ACActive: (callback) => {
+				if (!(device.state instanceof Classes.InternalAcState)) {
+					// TODO: log warning
+					return
+				}
+
 				const active = device.state.active
 				const mode = device.state.mode
 
 				if (!active || mode === 'FAN' || mode === 'DRY') {
-					log.easyDebug(device.name, '(GET) - AC Active State: false')
+					easyDebugInfo(device.name, '(GET) - AC Active State: false')
 
 					callback(null, 0)
 				} else {
-					log.easyDebug(device.name, '(GET) - AC Active State: true')
+					easyDebugInfo(device.name, '(GET) - AC Active State: true')
 
 					callback(null, 1)
 				}
 			},
 
+			/** @param {homebridge.CharacteristicGetCallback} callback */
 			PureActive: (callback) => {
+				if (!(device.state instanceof Classes.InternalAcState)) {
```

Comment:
```suggestion
				if (!(device.state instanceof Classes.InternalAirPurifierState)) {
```

Only called from AirPurifier